compiler/rustc_typeck/src/expr_use_visitor.rs

```diff
@@ -237,18 +238,26 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
                 // We only want to borrow discr if the pattern contain something other
                 // than wildcards
                 let ExprUseVisitor { ref mc, body_owner: _, delegate: _ } = *self;
-                let mut res = false;
+                let mut needs_to_be_read = false;
                 for arm in arms.iter() {
                     return_if_err!(mc.cat_pattern(discr_place.clone(), &arm.pat, |_place, pat| {
                         if let PatKind::Binding(_, _, _, opt_sub_pat) = pat.kind {
                             if let None = opt_sub_pat {
-                                res = true;
+                                needs_to_be_read = true;
+                            }
+                        } else if let PatKind::TupleStruct(qpath, _, _) = &pat.kind {
+                            // If a TupleStruct has a Some PathSegment, we should read the discr_place
+                            // regardless if it contains a Wild pattern later
```

Comment:
I think I would find this comment hard to understand if I lacked context. Can you add a code example?

Reply:
I think it would be great to e.g. showing matching on an `Option` and explain that this does require reading the discriminant, even if we don't load any data.